staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -113,71 +114,130 @@ func WithPriorityAndFairness(
-		// TODO: Estimate cost should also take fcIfc.GetWatchCount(requestInfo) as a parameter.
-		width := widthEstimator.EstimateWidth(r)
-		digest := utilflowcontrol.RequestDigest{RequestInfo: requestInfo, User: user, Width: width}
+			// Ensure that an item can be put to resultCh asynchronously.
+			resultCh := make(chan interface{}, 1)
 
-		fcIfc.Handle(ctx, digest, note, func(inQueue bool) {
-			if inQueue {
-				noteWaitingDelta(1)
-			} else {
-				noteWaitingDelta(-1)
+			// Call Handle in a separate goroutine.
+			// The reason for it is that from APF point of view, the request processing
+			// finishes as soon as watch is initialized (which is generally orders of
+			// magnitude faster then the watch request itself). This means that Handle()
+			// call finishes much faster and for performance reasons we want to reduce
+			// the number of running goroutines - so we run the shorter thing in a
+			// dedicated goroutine and the actual watch handler in the main one.
+			go func() {
+				defer func() {
+					err := recover()
+					// do not wrap the sentinel ErrAbortHandler panic value
+					if err != nil && err != http.ErrAbortHandler {
+						// Same as stdlib http server code. Manually allocate stack
+						// trace buffer size to prevent excessively large logs
+						const size = 64 << 10
+						buf := make([]byte, size)
+						buf = buf[:runtime.Stack(buf, false)]
+						err = fmt.Sprintf("%v\n%s", err, buf)
+					}
+
+					// Ensure that the result is put into resultCh independently of the panic.
+					resultCh <- err
+				}()
+
+				// We create handleCtx with explicit cancelation function.
+				// The reason for it is that Handle() underneath may start additional goroutine
+				// that is blocked on context cancellation. However, from APF point of view,
+				// we don't want to wait until the whole watch request is processed (which is
+				// when it context is actually cancelled) - we want to unblock the goroutine as
+				// soon as the request is processed from the APF point of view.
+				//
+				// Note that we explicitly do NOT call the actuall handler using that context
+				// to avoid cancelling request too early.
+				handleCtx, handleCtxCancel := context.WithCancel(ctx)
+
+				fcIfc.Handle(handleCtx, digest, note, queueNote, execute)
+				handleCtxCancel()
+			}()
+
+			defer func() {
+				if err := recover(); err != nil {
+					// In case of panic from inline handler, ensure that we clean up after
+					// ourselves, in particular unblock hanging goroutines.
+					defer panic(err)
+				}
+
+				// Protect from the situation when request will not reach storage layer
+				// and the initialization signal will not be send.
+				if watchInitializationSignal != nil {
+					watchInitializationSignal.Signal()
+				}
+				// Forget the watcher if it was registered.
+				if forgetWatch != nil {
+					forgetWatch()
+				}
+			}()
+
+			select {
+			case <-shouldStartWatchCh:
```

Comment:
if the goroutine executing `fcIfc.Handle` panics after we are here (it's an edge case) then the following block `err := <-resultCh:` never executes and we don't propagate the error to the caller go routine.

so the goroutine executing `fcIfc.Handle` can run for some time after `handler.ServeHTTP(w, watchReq)` starts executing.

maybe we can split it like below?
```
select {
			case <-shouldStartWatchCh:
				handler.ServeHTTP(w, watchReq)
				if err := <-resultCh; err != nil {
					panic(err)
				}
			case <-handleCtx.Done():
}

// the following channel should not block forever.
err := <-resultCh:
if err != nil {
    if served {
          // log the error
    } else {	
       panic(err)
   }
}
```

Reply:
oh, i missed the part where you read from `resultCh` and panic if there is an error
```
			case <-shouldStartWatchCh:
				handler.ServeHTTP(w, watchReq)
				if err := <-resultCh; err != nil {
					panic(err)
				}
```

I think this resolves the problem. The question is should we panic here or log the error? The watch request has completed after `handler.ServeHTTP(w, watchReq)` and the response has already been set appropriately. I think panic is the right approach, it makes the error more visible. 
Ideally we should terminate the request as soon as `fcIfc.Handle` panics, but this can be a follow up.

Reply:
Added a todo about termination.